### ISSUE TITLE
crouton on external SSD does not always allow clean unmount of external drive

### ISSUE THREAD

**Issue by robertstrom** [dnschneid/crouton#2314]
I have a crouton installed on an external SSD drive that works just fine while I am using it. My problem occurs when I go to log out / exit the trusty XFCE crouton I run into issues where the crouton does not appear to fully close out properly leaving me unable to properly eject the external drive (this also appears to leave the ChromeOS in an unstable state at times [the ChromeOS may crash unexpectedly after pulling the external drive]).

This issue does not happen all the time (it seems to occur approximately 50% of the time) but I'd rather it not happen at all.

Please let me know what information is needed to help diagnose this issue.

Thanks,

Robert

**Comment by DennisLfromGA** [comment]
It could be some parts/processes of the chroot are still busy & mounted.
Try using the 'unmount-chroot' command to find out and, if necessary, force the processes to terminate and unmount.

Something like: `sudo unmount-chroot -c /path/to/external/media/chroots -p`
will show you if anything is still busy and mounted.

Here's the usage text for 'unmount-chroot' -
```
Unmounts one or more chroots, optionally killing any processes still running
inside them.
By default, it will run in interactive mode where it will ask to kill any
remaining processes if unable to unmount the chroot within 5 seconds.
Options:
    -a          Unmount all chroots in the CHROOTS directory.
    -c CHROOTS  Directory the chroots are in. Default: $CHROOTS
    -f          Forces a chroot to unmount, potentially breaking or killing
                other instances of the same chroot.
    -k KILL     Send the processes SIGKILL instead of SIGTERM.
    -p          Print to STDOUT the processes stopping a chroot from unmounting.
    -t TRIES    Number of seconds to try before signalling the processes.
                Use -t inf to be exceedingly patient. Default: $TRIES
    -x          Keep the root directory of the chroot mounted.
    -y          Signal any remaining processes without confirmation.
                Automatically escalates from SIGTERM to SIGKILL."
```

**Comment by robertstrom** [comment]
I have used the unmount-chroot command several times with no luck. I have
also tried to figure out what process(es) may be hanging by using the
unmount-chroot -p command. I have also used the -f and -k switches with no
luck.

I only wish it was that easy ... ;-)

thanks

**Comment by DennisLfromGA** [comment]
@robertstrom,

> I have also tried to figure out what process(es) may be hanging by using the
unmount-chroot -p command. 
> I have also used the -f and -k switches with **no
luck**.

Does '*no luck*' mean the processes terminated but you still couldn't eject the media properly?
Or do you mean the processes did not terminate?

> I only wish it was that easy ... ;-)

Me too ;)

**Comment by robertstrom** [comment]
There were no processes listed / there were no processes to terminate and
the media never ejects cleanly whenever I use these commands. The external
drive is always still seen as busy by the ChromeOS.

**Comment by DennisLfromGA** [comment]
If there are no chroot processes listed then maybe it's not a crouton issue preventing it from ejecting, maybe it's a Chrome OS issue. I think as long as you're not writing to the media, it's safe to remove it.

**Comment by divx118** [comment]
@robertstrom Did you try ejecting it first in the File app of chromeos?

**Comment by divx118** [comment]
@robertstrom Did you try ejecting it first in the File app of chromeos?

**Comment by robertstrom** [comment]
Yes, every time

**Comment by robertstrom** [comment]
ChromeOS complains about it every time and there are many times when the OS
gets unstable (it ends up rebooting itself somewhere within the next hour
or so).

I think that there is a cyclical mount point issue of some sort. The depth
and repetitiveness of the mount does not look right to me. This is what the
results of the mount command look like an unsuccessful (external drive
still busy) logoff of the chroot looks like (I grepped for Chroots):

media on /media/removable/Chroots/chroot/xfce/var/host/media type tmpfs
(rw,nosuid,nodev,noexec,relatime)
/dev/sdd1 on
/media/removable/Chroots/chroot/xfce/var/host/media/removable/Chroots type
ext4 (rw,nosuid,nodev,noexec,relatime,dirsync)
none on /media/removable/Chroots/chroot/xfce/sys type sysfs
(rw,nosuid,nodev,noexec,relatime)
debugfs on /media/removable/Chroots/chroot/xfce/sys/kernel/debug type
debugfs (rw,nosuid,nodev,noexec,relatime,gid=236,mode=750)
none on /media/removable/Chroots/chroot/xfce/sys/fs/cgroup type tmpfs
(rw,nosuid,nodev,noexec,relatime,mode=755)
cgroup on /media/removable/Chroots/chroot/xfce/sys/fs/cgroup/cpu type
cgroup (rw,nosuid,nodev,noexec,relatime,cpu)
none on
/media/removable/Chroots/chroot/xfce/var/host/media/removable/Chroots/chroot/xfce/sys
type sysfs (rw,nosuid,nodev,noexec,relatime)
debugfs on
/media/removable/Chroots/chroot/xfce/var/host/media/removable/Chroots/chroot/xfce/sys/kernel/debug
type debugfs (rw,nosuid,nodev,noexec,relatime,gid=236,mode=750)
none on
/media/removable/Chroots/chroot/xfce/var/host/media/removable/Chroots/chroot/xfce/sys/fs/cgroup
type tmpfs (rw,nosuid,nodev,noexec,relatime,mode=755)
cgroup on
/media/removable/Chroots/chroot/xfce/var/host/media/removable/Chroots/chroot/xfce/sys/fs/cgroup/cpu
type cgroup (rw,nosuid,nodev,noexec,relatime,cpu)

**Comment by robertstrom** [comment]
Anybody have a chance to take a look at the above results of the mount command after an unsuccessful chroot shutdown (trusty XFCE shutdown but the external disk will not eject / external drive is still busy)?

Any thoughts / ideas?

Thanks,

Robert

**Comment by DennisLfromGA** [comment]
@robertstrom,

I'm a bit confused, you said here (https://github.com/dnschneid/crouton/issues/2314#issuecomment-167654762) that there were no processes listed to terminate but in this post (https://github.com/dnschneid/crouton/issues/2314#issuecomment-167669888) you listed several processes after logging out of your chroot.

I think the '`unmount-chroot`' command that I mentioned here (https://github.com/dnschneid/crouton/issues/2314#issuecomment-167467620) will be your best bet.
I don't know why they aren't terminating, I believe it's becasue they're on external media - just a guess though.

**Comment by robertstrom** [comment]
@DennisLfromGA 

Thanks for the reply ...

You are correct, I did say that there were no processes listed to terminate. That statement was based on the results of the unmount-chroot -p command, the - p is supposed to print to STDOUT the processes stopping a chroot from unmounting.

I did indeed get zero processes listed when I ran that command (which I have done numerous times).

The results that I posted were from the mount command with a grep filter of **Chroots**.

So ... the unmount-chroot command has not worked for me ... I have tried it MANY times.

I don't know why the processes aren't terminating either, hence the reason for my post. I'm sure that the cause is tied to the fact that the chroot is on the external media too but we're supposed to be able to move the chroot to external media and I have seen many posts of people doing so.

I have tried running the chroot from a few different types of external devices with the same results all on a Toshiba CB35-B3340.

First I tried running it from a fast MicroSD card in an SD adapter. While the MicroSD is fast enough it appears that the SD Card slot itself does not support any fast transfers so that possibility was immediately ruled out (it would have been the best since the card is entirely housed in the computer and could just stay there).

Then I tried a SanDisk Extreme CZ80 64GB USB 3.0 Flash Drive using the USB 3 slot on the Chromebook - http://www.amazon.com/gp/product/B00KT7DOSE?psc=1&redirect=true&ref_=oh_aui_search_detailpage - it is fast and pretty darn inexpensive. All worked just fine except for the problems with not properly ejecting.

The most recent attempt is with the chroot running on a Crucial MX200 500GB SATA 2.5 Inch Internal Solid State Drive using the USB 3 slot on the Chromebook - http://www.amazon.com/gp/product/B00RQA6E20?psc=1&redirect=true&ref_=oh_aui_search_detailpage - again, this external media is plenty fast enough  (Sequential reads/writes up to 555 / 500 MB/s on all file types) but the same media eject issues persist.

In summary, everything that I've read suggests that running chroots on external drives is supported / works. It sure would be nice if I could properly eject the media all the time (it doesn't always fail, only about 50% of the time).

Hopefully someone out there who works on this project will take a look at this and get this issue fixed.

I'll be happy to provide any information to help resolve this issue.

Thanks!

**Comment by robertstrom** [comment]
Hoping that @dnschneid will take a look @ this issue.

Thanks!

**Comment by mikehgentry** [comment]
I'm seeing this too I believe, on a Dell CB 11 (3120, Candy), Stable channel.

I was using micro SD cards for my chroots, and didn't have this problem. I bought a Sandisk Ultra Fit 128GB USB 3.0 mini thumb drive, and chroots on this frequently exhibit it - particularly at the end of creating a new chroot.

If I run 'top' in another shell, I can see three processes seemingly hang in succession: 'mountpoint', 'sed' and 'mount'. It takes something like 5 - 10 minutes to get through from mountpoint to mount, and each process is stuck at ~100 percent of a core while it's running. Mount never finishes, but I can ctrl-c out of it (sometimes at least, possibly always - I can't remember).

Any attempt to run 'mount' after this has happened just freezes (the shell doesn't respond in any way, though again, I can ctrl-c out of it). Only recourse is to shut the machine down, which only makes it part of the way, and I have to hold the power button to force power off.

I don't know if any of that's helpful, but let me know if there's any more information I can give you or experiments I can run.
